File: backend/app/services/notify.py
"""Helper to write notifications from anywhere in the backend.

audience:
  'admin'  — only admin users see this
  'all'    — every logged-in user sees this
  'user'   — only the user with target_user_id sees this
"""
from app.core.database import SessionLocal
from app.models.notification import Notification


def push(
    type_: str,
    title: str,
    body: str = "",
    audience: str = "admin",
    target_user_id: int | None = None,
) -> None:
    db = SessionLocal()
    try:
        n = Notification(
            type=type_,
            title=title,
            body=body,
            audience=audience,
            target_user_id=target_user_id,
        )
        db.add(n)
        db.commit()
        # Keep at most 300 latest notifications
        ids = db.query(Notification.id).order_by(Notification.id.desc()).offset(300).all()
        if ids:
            db.query(Notification).filter(
                Notification.id.in_([r[0] for r in ids])
            ).delete(synchronize_session=False)
            db.commit()
    except Exception:
        db.rollback()
    finally:
        db.close()

File: backend-go/services/notify.go
package services

import (
	"time"

	"simnexus-go/database"
	"simnexus-go/models"
)

// Push inserts a notification row. audience: "admin"|"support"|"all"|"user".
// For "user" audience pass a non-nil targetUserID.
func Push(nType, title, body, audience string, targetUserID *uint) {
	if audience == "" {
		audience = "admin"
	}
	n := models.Notification{
		Type:         nType,
		Title:        title,
		Body:         body,
		Audience:     audience,
		TargetUserID: targetUserID,
		CreatedAt:    time.Now(),
	}
	database.DB.Create(&n)
}
